Add basic hashing API

imalsogreg opened the pull request:
- [x] Internal module for data definitions and implementation
  - Everything is exported from `.Internal.Hash` ( "internal-by-convention" for unsafe constructors and functions )
 - [x] `System.Nix.Hash` reexports only the public-facing parts of `.Internal.Hash`
 - [x] `initialize`/`update`/`finalize` available for MD5, SHA-1, SHA-256
 - [x] `hash`/`hashLazy` functions provided for hashing bytestrings in one go (as in cryptonite)
 - [x] Tests (We have 2 tiny unit tests, one for `SHA1 -> base32`, one for `SHA256 -> truncated base32`)

shlevy commented on hnix-store-core/src/System/Nix/Internal/Hash.hs:
```diff
@@ -0,0 +1,152 @@
+  type AlgoCtx a :: *
+  hashLength        :: Proxy a -> Int
+
+  validateHashBytes :: BS.ByteString -> Maybe (Digest a)
```
Hmm... Do we need this? We might, but my inclination is to not expose it until we have a use case.

shlevy commented on hnix-store-core/src/System/Nix/Internal/Hash.hs:
```diff
@@ -0,0 +1,152 @@
+  update            :: AlgoCtx a -> BS.ByteString -> AlgoCtx a
+  finalize          :: AlgoCtx a -> Digest a
+
+  toBase32          :: Digest a -> T.Text
```
I don't think this is a class method really... Should just be a method on `Digest a`s

imalsogreg replied:
I'm using the class signal different types going to base32 in different ways (where MD5 and SHA1 simply base32Bytes their hash, while a SHA256 hash needs to do truncation first).

The fact that different types have different `base32` behaviors in my code might be a mistake in my mental model. Like the way I don't have a `Truncated` newtype in the API. If we _did_ have `newtype Truncated = T BS.Bytestring`, then I could see defining a few monomorphic funuctions `base32FromMD5` `base32FromTruncatedSHA256`, `base32FromSHA1`. A polymorphic `base32` function over `Digest a` would be able to produce some strings that aren't used by nix - like `base32 (someHash :: Digest 'SHA256)` <- that would be an "invalid" (not used by nix) `Text`.

imalsogreg replied:
Thesis (Page 95) talks about the `valid : Path -> Hash` mapping, which does use untruncated SHA256 hashes. So yep, have to rethink this :)

shlevy commented on hnix-store-core/src/System/Nix/Internal/Hash.hs:
```diff
@@ -0,0 +1,152 @@
+class HasDigest (a :: HashAlgorithm) where
+
+  type AlgoCtx a :: *
+  hashLength        :: Proxy a -> Int
```
nit: `proxy a -> Int`.

FWIW I prefer [Tagged](http://hackage.haskell.org/package/tagged-0.8.6/docs/Data-Tagged.html), but I know conventions are against me here

shlevy commented on hnix-store-core/src/System/Nix/Internal/Hash.hs:
```diff
@@ -0,0 +1,152 @@
+  update = SHA256.update
+  finalize = Digest . SHA256.finalize
+  toBase32 (Digest bs) =
+    T.decodeUtf8 $ printHashBytes32 $ truncateDigest 20 bs
```
? We definitely need untruncated sha-256 hashes (and their base 32 representations) as well as truncated ones, and in any case if we're dealing with truncated hashes it's the entire hash that's truncated, not just its human-friendly representation (e.g. when comparing for equality). IMO `HashAlgorithm` should have a `Truncated` constructor taking another `HashAlgorithm`

imalsogreg replied:
```
let h_bytes = hash256(string)
    h_bytes' = truncate(h_bytes)
    forStorePath = toBase32(h_bytes')
```
Imagining I go through these steps to produce a store path hash. There are multiple places in the code where we need the untruncated digest, and multiple places where we need the final base32 encoded truncated hash.

Is there any place in the code other than generating base32 encodings, where we would need access to the truncated version of the hash bytes? If so, then a `Truncated` newtype would be useful for carrying the proof that truncation has happened. OTOH if truncation is merely an implementation detail of producing base32 encoding of a hash, then `Truncated` newtype wouldn't pull its weight. What do you think?

shlevy commented on hnix-store-core/src/System/Nix/Internal/Hash.hs:
```diff
@@ -0,0 +1,152 @@
+
+-- TODO Test this, too
+truncateDigest :: Int -> BS.ByteString -> BS.ByteString
+truncateDigest n c = BS.pack $ map truncOutputByte [0.. n-1]
```
Why isn't this just `BS.take`?

imalsogreg replied:
This is a straight (and inefficient!) translation of the thesis. Because `n < |c|` (truncated length is less than the input bytestring length), some positions in the input bytestring get to impact the truncated version twice.

A more straightforward implementation would be: 
```
let layer1  = BS.take n bs
    -- leading bits that aren't truncated
    layer2 = BS.drop n bs <> BS.replicate (BS.length bs - n) 0
    -- ^ trailing bits, right padded
    -- v xor the two together for "truncated" digest
in xor layer1 layer2
``` (if xor could work bitwise over a bytestring)

shlevy replied:
Wait, seriously? It's not just a naive truncation? Why???

shlevy replied:
```c++
    Hash h;
    h.hashSize = newSize;
    for (unsigned int i = 0; i < hash.hashSize; ++i)
        h.hash[i % newSize] ^= hash.hash[i];
    return h;
```
OK then

imalsogreg commented:
@shlevy, Addressed all your comments. Especially the one about moving `base32` out of the typeclass, since yes, non-truncated sha256 hashes do sometimes get used in nix-store.

No hurry, just FYI. @sorki, if you have time, would love to get your review too!

shlevy commented on hnix-store-core/src/System/Nix/Internal/Hash.hs:
```diff
@@ -0,0 +1,159 @@
+import qualified Data.ByteString.Lazy  as BSL
+import qualified Data.Hashable         as DataHashable
+import           Data.List             (foldl')
+import           Data.Proxy            (Proxy(Proxy))
```
Unnecessary import?

imalsogreg replied:
At the time it was leftover. Now it's needed for a use of `natVal`.

shlevy commented:
(I suppose there could possibly be some case where we want user code to have a `Digest a -> Digest (Truncated n a)` to avoid recalculating the hash, but that's a pure optimization that crosses conceptual boundaries)

imalsogreg commented:
I haven't been thinking of truncation as a type of hashing algorithm. In nix world, SHA256 is the only algo whose digest ever gets truncated, so I wonder if we should enforce that as: 

```
data HashAlgorithm =
  = MD5
  | SHA1
  | SHA256
  | TruncatedSHA256 Nat
``` 
?
Furthermore, `20` is the only truncation length ever used in nix. So, drop that type param too? Or should we leave it polymorphic in case _we_ ever want a larger family or algorithm tags?

Re: sometimes wanting a truncated digest and sometimes wanting an untruncated digest, here is an example. The top is part of some validity checking, where a FSO may be read and hashed without truncation. The bottom part is computing a store hash, where of course we _do_ truncate.

```
valid[p] ← "sha256:"+printHash32(hash_sha256(serialise(readPath(p))))
```
vs.

```
makePath(type, descr, name) =
nixStore+"/"+printHash32(truncate(20, hash_sha256(s))) +"-"+name
where
s = type+":sha256:"+descr+":"+nixStore+":"+name
```

shlevy commented:
> In nix world, SHA256 is the only algo whose digest ever gets truncated

`nix hash` can `--truncate` any hash, though it's true for the actual store interface we are only ever truncating sha256.

> Furthermore, 20 is the only truncation length ever used in nix. So, drop that type param too? Or should we leave it polymorphic in case we ever want a larger family or algorithm tags?

> If the type-level nats are too painful, we can specialize this to `Truncated160` for now.

:wink: 

> Re: sometimes wanting a truncated digest and sometimes wanting an untruncated digest, here is an example.

Let's look at the actual store path hashing code:

```C++
Path Store::makeStorePath(const string & type,
    const Hash & hash, const string & name) const
{
    /* e.g., "source:sha256:1abc...:/nix/store:foo.tar.gz" */
    string s = type + ":" + hash.to_string(Base16) + ":" + storeDir + ":" + name;

    checkStoreName(name);

    return storeDir + "/"
        + compressHash(hashString(htSHA256, s), 20).to_string(Base32, false)
        + "-" + name;
}
```

It's true here we're passing in a hash, and then computing a truncated hash. But *the truncated hash is not a truncation of the passed in hash*. Instead, the truncated hash is the truncated hash of a string which includes as one of its parts the hex-encoded representation of the input hash. The input hash can even be a completely different hash algorithm, e.g. if you have a fixed output derivation and provide a `sha512` instead of a `sha256`.
